run/opencl/ed25519-donna/curve25519-donna-32bit.h

```diff
@@ -322,7 +322,7 @@ curve25519_square(bignum25519 out, const bignum25519 in) {
 }
 
 static void
-curve25519_mul_const(bignum25519 out, const bignum25519 a, __constant bignum25519 b) {
+curve25519_mul_const(bignum25519 out, const bignum25519 a, __constant uint32_t *b) {
```

Comment:
BTW, in one case I introduce a separate type with `_p` as suffix for a similar purpose:

```c
typedef unsigned char hash_512bits[64];
typedef unsigned char *hash_512bits_p;
```

In another case though, I reused the original type name:

```c
//typedef unsigned char ed25519_secret_key[32];
typedef unsigned char *ed25519_secret_key;
```

So we now have a mix of 3 approaches to the same porting task.

Reply:
Actually I first did `typedef uint32_t *bignum25519_p;` but somehow that didn't work.
```
Build log: In file included from <program source>:194:
In file included from opencl/ed25519-donna/ed25519-donna.c:7:
In file included from opencl/ed25519-donna/ed25519-donna.h:13:
opencl/ed25519-donna/curve25519-donna-32bit.h:326:85: error: parameter may not be qualified with an address space
curve25519_mul_const(bignum25519 out, const bignum25519 a, __constant bignum25519_p b) {
                                                                                    ^
```
To get back some readability, I could do something like `typedef uint32_t bignum25519_e;` (e for element?) and use `__constant bignum25519_e *b` in the function argument.

Reply:
You do use `ed25519_secret_key` and `hash_512bits_p` in function arguments, but only as private memory. Not sure why my compiler breaks when I use `__constant bignum25519_p b` - that detail might be an Apple Problem[tm]

Reply:
This works
```diff
diff --git a/run/opencl/ed25519-donna/curve25519-donna-32bit.h b/run/opencl/ed25519-donna/curve25519-donna-32bit.h
index 9c0c3bd79..25dee5541 100644
--- a/run/opencl/ed25519-donna/curve25519-donna-32bit.h
+++ b/run/opencl/ed25519-donna/curve25519-donna-32bit.h
@@ -6,6 +6,8 @@
 */
 
 typedef uint32_t bignum25519[10];
+//typedef uint32_t *bignum25519_p; /* Somehow this typedef doesn't work correctly on OSX */
+#define bignum25519_p uint32_t*
 typedef uint32_t bignum25519align16[12];
 
 #define reduce_mask_25 ((1 << 25) - 1)
@@ -322,7 +324,7 @@ curve25519_square(bignum25519 out, const bignum25519 in) {
 }
 
 static void
-curve25519_mul_const(bignum25519 out, const bignum25519 a, __constant uint32_t *b) {
+curve25519_mul_const(bignum25519 out, const bignum25519 a, __constant bignum25519_p b) {
        bignum25519 bpvt;
        memcpy_cp(bpvt, b, sizeof(bpvt));
        curve25519_mul(out, a, bpvt);
```

Reply:
Well, I have no preference here - none of these are ideal. We can as well leave it as it's in your currently pushed commits.